cms/djangoapps/contentstore/helpers.py

```diff
@@ -428,14 +430,29 @@ def _import_xml_node_to_parent(
                 tags=tags,
             )
 
+    def copy_tags():
+        if copied_from_block and tags:
+            object_tags = tags.get(str(copied_from_block))
+            if object_tags:
+                content_tagging_api.set_all_object_tags(
+                    content_key=new_xblock.location,
+                    object_tags=object_tags,
+                )
+
     # Copy content tags to the new xblock
-    if copied_from_block and tags:
-        object_tags = tags.get(str(copied_from_block))
-        if object_tags:
-            content_tagging_api.set_all_object_tags(
-                content_key=new_xblock.location,
-                object_tags=object_tags,
+    if new_xblock.upstream:
+        # Verify if the upstream is a library component
+        # Copy the tags from library component upstream as ready only
+        try:
+            LibraryUsageLocatorV2.from_string(new_xblock.upstream)
+            content_tagging_api.copy_tags_as_read_only(
+                new_xblock.upstream,
+                new_xblock.location,
             )
+        except InvalidKeyError:
+            copy_tags()
```

Comment:
```suggestion
        # If this block is synced from an upstream (e.g. library content),
        # copy the tags from the upstream as ready-only
        content_tagging_api.copy_tags_as_read_only(
            new_xblock.upstream,
            new_xblock.location,
        )
```
I think that we should treat all upstreams the same. It doesn't really matter for now though as we're only supporting libraries as upstreams. CC @kdmccormick in case you have a different opinion.

Reply:
Updated: https://github.com/openedx/edx-platform/pull/35596/commits/845f124a85a59ca34cde37bf36c94656cdc0c90f